### ISSUE TITLE
The directory walker stops functioning if --target-dir is passed to cargo

### ISSUE THREAD

**Issue by lu-zero** [jamesmunns/toml-cfg#7]
Se seen [here](https://github.com/jamesmunns/toml-cfg/blob/main/src/lib.rs#L244)
``` rust
    let mut out_dir = PathBuf::from(out_dir?);
    while !out_dir.ends_with("target") {
        if !out_dir.pop() {
            // We ran out of directories...
            return None;
        }
    }
```

in `pkg-example/application`
```console
$ cargo run --target-dir foo
...
32
hello
$ cargo run 
...
4096
Guten tag!
```

**Comment by jamesmunns** [comment]
Yes, I am aware of this and if there is a way to fix it, I'm open to supporting it. I don't know how, which is why it is like this.

Likely anything that changes the target path out of a recursive parent as well.

**Comment by jamesmunns** [comment]
This *might* be coupled to #3 as well

**Comment by lu-zero** [comment]
The two ways I was thinking were:
- if we can rely on cargo-metadata we can "simply" recover the information from cargo itself, but it is a bit ugly
- otherwise we can just walk up till we hit `/` but we might hit a false positive maybe?

**Comment by jamesmunns** [comment]
I'm open to it!

It would be good to have a test (suite) in the repo that exercises different cases. I think starting with one that:

* shows the current example working
* shows this issue failing (today)

So we can try code changes objectively, would be really neat.

**Comment by mvirkkunen** [comment]
This is made worse by the fact that Cargo now allows setting a global build directory which it used for all the temporary build files:

```
matti@aoba:~$ cat .cargo/config.toml 
[build]
build-dir = "/home/matti/.cargo/build"
```

This makes `--out-dir` point to something under that directory, completely breaking the walking. I use this to avoid having really big target directories all over my home directory, having to delete just one directory to get rid of the gigabytes of bloat is much more convenient.

I dumped all the environment variables and arguments and as far as I can tell, there really isn't any way to determine a "root crate" (the one passed to `cargo` initially) in this case. Only cargo knows, and I'm not sure how interested they would be in adding a new environment variable for us.

I can see two approaches:
- Start from `CARGO_MANIFEST_DIR` and walk up (to `/` in the end), but this feels ugly and subject to break in other ways - this would work for the usual "config file and crate in same workspace" case
- Add a way to pass in the path to the desired `cfg.toml`, e.g. via a custom environment variable
